### PR TITLE
EPNG-2521 tracing the suspect - a microservices murder mystery

### PR DESCRIPTION
**Pull request by jensfischerhh**

### CODE REVIEW

File: _posts/2017/2017-08-17-tracing-the-suspect.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
+---
+layout: post
+title: "Tracing the suspect - a microservices murder mystery"
+date: "2017-08-17 09:30:00"
+image: blog-header/tracing-the-suspect.jpg
```
Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: What about one of those?

https://pixabay.com/en/snow-lane-traces-snow-zig-zag-81156/
https://www.pexels.com/photo/cold-dark-eerie-fear-207985/
https://pixabay.com/en/sand-footsteps-footprints-beach-768783/

File: _posts/2017/2017-08-17-tracing-the-suspect.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+authors: ["Jens"]
+---
+
+The [first post][haystack] in this series of blog posts about debugging microservices covered, how we can produce structured log events to ease log aggregation.
+We also introduced the concept of log correlation using a `correlation-id`, that spans multiple microservices to connect related log events.
+In this second post we will enhance the JSON log structure, and dive deeper into the topic of **distributed tracing**.
+
+## Custom JSON structure
+
+Since at ePages we use [Google's Stackdriver Logging infrastructure][stackdriver] to aggregate all our log events, we need to adjust our JSON structure to follow their format:
```
Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: At ePages we use...., thus we need to...

File: _posts/2017/2017-08-17-tracing-the-suspect.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+We want to rename the default JSON properties `@timestamp` and `level` to `time` and `severity` respectively, while completely dropping `@version`, `level_value` and `thread_name`.
```
Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: ...`level_value`, and ... (add comma)

File: _posts/2017/2017-08-17-tracing-the-suspect.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            </providers>
+        </encoder>
+    </appender>
+    <root level="INFO">
+        <appender-ref ref="STDOUT"/>
+    </root>
+</configuration>
+{% endhighlight %}
+
+The JSON property`app` is rendered by the `<context/>` JSON provider, which gets its value from the `<springProperty/>` element used to access the `spring.application.name` configuration value available for every Spring Boot app.
```
Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: not sure if there's a space missing between property and app?

File: _posts/2017/2017-08-17-tracing-the-suspect.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
+---
+layout: post
+title: "Tracing the suspect - a microservices murder mystery"
```
Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: Rename file name same as title: 
_posts/2017/2017-08-17-tracing-the-suspect-a-microservices-murder-mystery.md

File: _posts/2017/2017-08-17-tracing-the-suspect.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+}
+{% endhighlight %}
+
+
+## And then it happened!
+
+In our example it is possible, that one of the participating microservices is producing an error, causing the whole request spanning all microservices to fail.
+For investigating these kinds of failures it is important to also get access to the stacktrace providing detailed information.
+Stacktraces in Java can get pretty unwieldy and long, and they happen to hide the most important information at the very bottom.
+We introduce a special `<stackTrace/>` JSON provider, that can be configured to truncate noisy stacktrace frames (e.g. from invocation via reflection) and moves the root cause to the top:
```
Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: ... via reflection), and moves the... (add comma)

File: _posts/2017/2017-08-17-tracing-the-suspect.md (hunk 1 above)
Comment (jensfischerhh) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: 1st: difficult to understand it'S footprints
2nd: mysterious, but not yet convincing
3rd: good footprints visualisation, just missing some kind of crime scene analogy.

Really undecided on that topic, but not too keen on finding the perfect fit.

Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: Another bunch:

https://pixabay.com/en/night-luna-sky-full-moon-clouds-578091/
https://unsplash.com/collections/487337/murder-issue?photo=3jcc_SeZ_UA
https://unsplash.com/collections/1026362/crime
https://unsplash.com/collections/1026362/crime?photo=9PPvnEd92xE
https://unsplash.com/collections/1026362/crime?photo=8SSuaY2e_EY
https://unsplash.com/search/photos/footsteps?photo=83FN3i0cSNU

File: _posts/2017/2017-08-17-tracing-the-suspect.md [hunk 7]
```diff
@@ -2,7 +2,7 @@
 layout: post
 title: "Tracing the suspect - a microservices murder mystery"
 date: "2017-08-17 09:30:00"
-image: blog-header/tracing-the-suspect.jpg
+image: blog-header/tracing-the-suspect-a-microservices-murder-mystery.jpg
```
Comment (BirgitBader) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: sorry, that was a misunderstanging. I was not relating to the image but to the .md file name

File: _posts/2017/2017-08-17-tracing-the-suspect.md (hunk 1 above)
Comment (jensfischerhh) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: I like https://pixabay.com/en/night-luna-sky-full-moon-clouds-578091/ and https://unsplash.com/photos/9PPvnEd92xE. You make the final decision. Or just stick to the gun :-)

Comment (jensfischerhh) [inline _posts/2017/2017-08-17-tracing-the-suspect.md]: or this https://pixabay.com/en/alone-astronomy-blue-2297211/

**Pull request merged**